File: ios-native/Sources/ALUI/Components/EndButtons.swift
import SwiftUI

import ALCore

/* -------------------------------------------------------------------------- */
/* `src/components/EndButtons.tsx` — leave, or keep going.                     */
/*                                                                             */
/* ```tsx                                                                      */
/* <div className="mt-1 flex flex-wrap items-center justify-center gap-3">     */
/*   <button                                                                   */
/*     onPointerDown={onMenu}                                                  */
/*     className="rounded-full bg-white/80 px-7 py-4 text-xl font-extrabold     */
/*                text-[#5A3A1E] shadow [touch-action:none]"                   */
/*   >🏠 Menu</button>                                                          */
/*   <button                                                                   */
/*     onPointerDown={onNext}                                                  */
/*     className="rounded-full bg-[#66BB6A] px-9 py-4 text-2xl font-extrabold   */
/*                text-white [touch-action:none]"                              */
/*     style={{ boxShadow: "0 8px 0 #43A047, 0 14px 24px rgba(0,0,0,0.2)" }}   */
/*   >🎉 Suivant</button>                                                       */
/* </div>                                                                      */
/* ```                                                                         */
/*                                                                             */
/* Both buttons fire on `onPointerDown`, not on click — so both go through the */
/* app-wide `touchDown` primitive (D5, invariant 1). `[touch-action:none]` is  */
/* the web's way of saying "this gesture is mine, do not scroll"; the          */
/* recogniser handles that by construction.                                    */
/*                                                                             */
/* Neither button has a `font-black`: they are `font-extrabold` (800), which is */
/* SwiftUI `.heavy`, NOT `.black`. `Typography.Weight` spells that trap out.   */
/* -------------------------------------------------------------------------- */

public struct EndButtons: View {

    /// `gap-3`.
    public static let gap: CGFloat = 12
    /// `mt-1`.
    public static let topMargin: CGFloat = 4

    /// « 🏠 Menu » — `px-7 py-4 text-xl`, `bg-white/80`.
    public static let menuPaddingX: CGFloat = 28
    public static let menuPaddingY: CGFloat = 16

    /// « 🎉 Suivant » — `px-9 py-4 text-2xl`, `bg-[#66BB6A]`.
    public static let nextPaddingX: CGFloat = 36
    public static let nextPaddingY: CGFloat = 16

    /// `0 8px 0 #43A047`.
    public static let nextLipDrop: CGFloat = 8
    /// `0 14px 24px rgba(0,0,0,0.2)`.
    public static let nextSoftShadow = CSSShadow(y: 14, blur: 24, opacity: 0.2)

    /// Tailwind's plain `shadow` utility:
    /// `0 1px 3px 0 rgb(0 0 0/0.1), 0 1px 2px -1px rgb(0 0 0/0.1)`. The two
    /// layers are within a point of each other, so they collapse into one.
    public static let menuShadow = CSSShadow(y: 1, blur: 3, opacity: 0.1)

    public let onMenu: () -> Void
    public let onNext: () -> Void

    public init(onMenu: @escaping () -> Void, onNext: @escaping () -> Void) {
        self.onMenu = onMenu
        self.onNext = onNext
    }

    public var body: some View {
        ComponentsWrapRow(spacing: Self.gap, lineSpacing: Self.gap) {
            menuButton
            nextButton
        }
        .padding(.top, Self.topMargin)
    }

    private var menuButton: some View {
        Text(verbatim: Copy.EndButtons.menu)
            .font(Typography.rounded(Typography.Size.xl, Typography.Weight.extrabold))
            .foregroundStyle(Palette.ink.color)
            .padding(.horizontal, Self.menuPaddingX)
            .padding(.vertical, Self.menuPaddingY)
            .background {
                Capsule()
                    .fill(.white.opacity(Palette.White.o80))
                    .shadow(
                        color: .black.opacity(Self.menuShadow.opacity),
                        radius: Self.menuShadow.swiftUIRadius,
                        x: 0,
                        y: Self.menuShadow.y
                    )
            }
            .contentShape(Capsule())
            .touchDown { onMenu() }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(Text(verbatim: Copy.EndButtons.menu))
            .accessibilityAddTraits(.isButton)
    }

    private var nextButton: some View {
        Text(verbatim: Copy.EndButtons.next)
            .font(Typography.rounded(Typography.Size.xxl, Typography.Weight.extrabold))
            .foregroundStyle(.white)
            .padding(.horizontal, Self.nextPaddingX)
            .padding(.vertical, Self.nextPaddingY)
            .background {
                liftedCapsule(
                    fill: AnyShapeStyle(Palette.green.color),
                    lip: Palette.greenLip.color,
                    drop: Self.nextLipDrop,
                    soft: Self.nextSoftShadow
                )
            }
            .contentShape(Capsule())
            .touchDown { onNext() }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(Text(verbatim: Copy.EndButtons.next))
            .accessibilityAddTraits(.isButton)
    }
}

/* -------------------------------------------------------------------------- */
/* `flex flex-wrap items-center justify-center gap-N`.                         */
/*                                                                             */
/* SwiftUI has no wrapping stack, and both places this module needs one — the  */
/* end buttons and the Finished star recap — are authored `flex-wrap` in the   */
/* TSX. An `HStack` would instead SHRINK its children when the row runs out of */
/* room, which on a narrow phone squashes « 🎉 Suivant » rather than dropping  */
/* it to a second line. One `Layout`, used by both.                            */
/* -------------------------------------------------------------------------- */

// NB: qualified. `ALCore` exports a `Layout` type (the mascot growth model), so
// a bare `Layout` here is « ambiguous for type lookup ». Same trap as
// `ALArt/Mascot/Rig.swift`; `LayoutSubviews` is spelled out for the same reason.
struct ComponentsWrapRow: SwiftUI.Layout {
    /// `gap-N` along the main axis.
    var spacing: CGFloat
    /// `gap-N` between wrapped lines (CSS `gap` is both).
    var lineSpacing: CGFloat

    struct Line {
        var indices: [Int] = []
        var sizes: [CGSize] = []
        var width: CGFloat = 0
        var height: CGFloat = 0
    }

    func lines(_ subviews: LayoutSubviews, maxWidth: CGFloat) -> [Line] {
        let sizes = subviews.indices.map { subviews[$0].sizeThatFits(ProposedViewSize.unspecified) }
        return Self.lines(sizes: sizes, maxWidth: maxWidth, spacing: spacing)
    }

    /// The wrapping arithmetic, on measured sizes alone — `LayoutSubviews`
    /// cannot be constructed outside a running layout pass, and this is the
    /// part that has to be right.
    static func lines(sizes: [CGSize], maxWidth: CGFloat, spacing: CGFloat) -> [Line] {
        var out: [Line] = []
        var current = Line()
        for index in sizes.indices {
            let size = sizes[index]
            if current.indices.isEmpty {
                current.indices = [index]
                current.sizes = [size]
                current.width = size.width
                current.height = size.height
                continue
            }
            let widened = current.width + spacing + size.width
            if widened > maxWidth {
                out.append(current)
                current = Line(indices: [index], sizes: [size], width: size.width, height: size.height)
            } else {
                current.indices.append(index)
                current.sizes.append(size)
                current.width = widened
                current.height = Swift.max(current.height, size.height)
            }
        }
        if !current.indices.isEmpty { out.append(current) }
        return out
    }

    func sizeThatFits(
        proposal: ProposedViewSize,
        subviews: LayoutSubviews,
        cache: inout ()
    ) -> CGSize {
        let limit: CGFloat = proposal.width ?? .infinity
        let rows = lines(subviews, maxWidth: limit)
        var width: CGFloat = 0
        var height: CGFloat = 0
        for row in rows {
            width = Swift.max(width, row.width)
            height += row.height
        }
        let gaps: CGFloat = CGFloat(Swift.max(0, rows.count - 1)) * lineSpacing
        return CGSize(width: width, height: height + gaps)
    }

    func placeSubviews(
        in bounds: CGRect,
        proposal: ProposedViewSize,
        subviews: LayoutSubviews,
        cache: inout ()
    ) {
        let rows = lines(subviews, maxWidth: bounds.width)
        var y = bounds.minY
        for row in rows {
            // justify-center
            var x = bounds.minX + (bounds.width - row.width) / 2
            for (slot, index) in row.indices.enumerated() {
                let size = row.sizes[slot]
                subviews[index].place(
                    at: CGPoint(x: x, y: y + row.height / 2),   // items-center
                    anchor: UnitPoint.leading,
                    proposal: ProposedViewSize(size)
                )
                x += size.width + spacing
            }
            y += row.height + lineSpacing
        }
    }
}

File: apps/game-android/ui/src/main/kotlin/fr/dappit/attrapelettres/ui/components/EndButtons.kt
package fr.dappit.attrapelettres.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import fr.dappit.attrapelettres.ui.design.Copy
import fr.dappit.attrapelettres.ui.design.Palette
import fr.dappit.attrapelettres.ui.design.Typography
import fr.dappit.attrapelettres.ui.interaction.touchDown

// Port of `src/components/EndButtons.tsx` — leave, or keep going.
//
//   <div className="mt-1 flex flex-wrap items-center justify-center gap-3">
//     <button
//       onPointerDown={onMenu}
//       className="rounded-full bg-white/80 px-7 py-4 text-xl font-extrabold
//                  text-[#5A3A1E] shadow [touch-action:none]"
//     >🏠 Menu</button>
//     <button
//       onPointerDown={onNext}
//       className="rounded-full bg-[#66BB6A] px-9 py-4 text-2xl font-extrabold
//                  text-white [touch-action:none]"
//       style={{ boxShadow: "0 8px 0 #43A047, 0 14px 24px rgba(0,0,0,0.2)" }}
//     >🎉 Suivant</button>
//   </div>
//
// Both buttons fire on `onPointerDown`, not on click, so both go through the
// app-wide `touchDown` primitive (invariant 1). `[touch-action:none]` is the
// web's way of saying « this gesture is mine, do not scroll »; the Compose
// recogniser takes the down on `PointerEventPass.Initial` and treats a consumed
// move as a cancel, which is the same contract.
//
// Neither button is `font-black`: they are `font-extrabold` (800). The gap
// between 800 and 900 is small and visible, and getting it wrong here would
// quietly make the end screen louder than the hub.
//
// Tailwind spacing is n x 4 px: mt-1 = 4, gap-3 = 12, py-4 = 16, px-7 = 28,
// px-9 = 36.

object EndButtonsMetrics {
    /** `gap-3`, on both axes — the row wraps on a narrow phone. */
    val GAP: Dp = 12.dp

    /** `mt-1`. */
    val TOP_MARGIN: Dp = 4.dp

    /** « 🏠 Menu » — `px-7 py-4 text-xl`, `bg-white/80`. */
    val MENU_PADDING_X: Dp = 28.dp
    val MENU_PADDING_Y: Dp = 16.dp

    /** « 🎉 Suivant » — `px-9 py-4 text-2xl`, `bg-[#66BB6A]`. */
    val NEXT_PADDING_X: Dp = 36.dp
    val NEXT_PADDING_Y: Dp = 16.dp

    /** `0 8px 0 #43A047`. */
    val NEXT_LIP_DROP: Dp = 8.dp

    /** `0 14px 24px rgba(0,0,0,0.2)`. */
    val NEXT_SOFT_SHADOW = CssShadow(y = 14.dp, blur = 24.dp, opacity = 0.2f)

    /**
     * Tailwind's plain `shadow` utility, both layers:
     * `0 1px 3px 0 rgb(0 0 0/0.1), 0 1px 2px -1px rgb(0 0 0/0.1)`.
     *
     * Taken from the shared [Shadows] table rather than re-authored — the same
     * utility is on the shop's back button and the trial chip, and two copies of
     * a Tailwind default is how they drift.
     */
    val MENU_SHADOW: List<CssShadow> = Shadows.tailwind
}

/**
 * End-of-run choices, shared by every exercise's finished screen: leave to the
 * hub, or keep going to the next level. « Suivant » is the primary (green)
 * button.
 */
@Composable
fun EndButtons(
    onMenu: () -> Unit,
    onNext: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val fontScale = LocalDensity.current.fontScale

    WrapRow(
        spacing = EndButtonsMetrics.GAP,
        modifier = modifier.padding(top = EndButtonsMetrics.TOP_MARGIN),
    ) {
        BasicText(
            text = Copy.EndButtons.MENU,
            style = Typography.style(
                size = Typography.Size.xl,
                weight = Typography.Weight.extrabold,
                color = Palette.ink.color,
                fontScale = fontScale,
            ),
            modifier = Modifier
                .cssShadow(EndButtonsMetrics.MENU_SHADOW, CircleShape)
                .background(Color.White.copy(alpha = Palette.White.o80), CircleShape)
                .touchDown { onMenu() }
                .padding(
                    horizontal = EndButtonsMetrics.MENU_PADDING_X,
                    vertical = EndButtonsMetrics.MENU_PADDING_Y,
                )
                .clearAndSetSemantics {
                    contentDescription = Copy.EndButtons.MENU
                    role = Role.Button
                },
        )

        BasicText(
            text = Copy.EndButtons.NEXT,
            style = Typography.style(
                size = Typography.Size.xxl,
                weight = Typography.Weight.extrabold,
                color = Color.White,
                fontScale = fontScale,
            ),
            modifier = Modifier
                .liftedPill(
                    fill = { SolidColor(Palette.green.color) },
                    lip = Palette.greenLip.color,
                    drop = EndButtonsMetrics.NEXT_LIP_DROP,
                    soft = EndButtonsMetrics.NEXT_SOFT_SHADOW,
                )
                .touchDown { onNext() }
                .padding(
                    horizontal = EndButtonsMetrics.NEXT_PADDING_X,
                    vertical = EndButtonsMetrics.NEXT_PADDING_Y,
                )
                .clearAndSetSemantics {
                    contentDescription = Copy.EndButtons.NEXT
                    role = Role.Button
                },
        )
    }
}
